tests/functional/test_metamodel/test_model_kwargs.py

```diff
@@ -0,0 +1,94 @@
+from __future__ import unicode_literals
+from textx import (metamodel_from_str)
+import os.path
+from pytest import raises
+from textx.exceptions import TextXError
+
+
+grammar = r"""
+Model: 'MyModel' name=ID;
+"""
+
+text = r"""
+MyModel test1
+"""
+
+
+def test_model_kwargs():
+    mm = metamodel_from_str(grammar)
+    mm._tx_model_kwarg_definitions.add_definition(
+        "parameter1", "an example param (1)"
+    )
+    mm._tx_model_kwarg_definitions.add_definition(
+        "parameter2", "an example param (2)"
+    )
+
+    m = mm.model_from_str(text, parameter1='P1', parameter2='P2')
+    assert m.name == 'test1'
+    assert hasattr(m, '_tx_model_kwargs')
+    assert len(m._tx_model_kwargs) == 2
+    assert len(m._tx_model_kwargs.used_keys) == 0
+
+    assert not m._tx_model_kwargs.have_all_parameters_been_used()
```

Comment:
`have_all_parameters_been_used` can be called just `all_been_used` and that could be Python property so no need for `()`

BTW, do we need checking if params has been used? We should make those params fully optional and the language designer should be aware that some/all of them might not be defined during model loading in which case a sane default should be used.

Reply:
"have been used" --> "have been used by the metamodel" and not "have been specified by the user"...

I am not sure if we need this. It could be a hint, that you accidentally provided an argument not used by your meta model... 

What do you think... It is easy to remove this option. In that case we could also use a simple dict instead of the ModelParams class.

Reply:
(added property).

Reply:
I think we should leave this option: if an option passed by the user has not been used, this is at least a warning in some cases... (maybe checked by a model processor).